packages/core/src/tracing/instrumentation/frameworks/fastify.js

```diff
@@ -47,51 +50,36 @@ function instrument(build) {
   function overwrittenBuild() {
     const app = build.apply(this, arguments);
 
-    if (app.route) {
-      const originalRoute = app.route;
-      app.route = function shimmedRoute(opts) {
-        if (opts.handler) {
-          const originalHandler = opts.handler;
-          opts.handler = function shimmedHandler() {
-            annotateHttpEntrySpanWithPathTemplate(app, opts);
-            return originalHandler.apply(this, arguments);
-          };
-        }
-
-        let preHandler;
-        let preHandlerKey;
-        if (opts.preHandler) {
-          // In Fastify 2.x, the attribute is called preHandler.
-          preHandler = opts.preHandler;
-          preHandlerKey = 'preHandler';
-        } else if (opts.beforeHandler) {
-          // In Fastify 1.x, the attribute is called beforeHandler.
-          preHandler = opts.beforeHandler;
-          preHandlerKey = 'beforeHandler';
-        }
-
-        if (preHandler) {
-          if (typeof preHandler === 'function') {
-            opts[preHandlerKey] = function shimmedPreHandler() {
-              annotateHttpEntrySpanWithPathTemplate(app, opts);
-              return preHandler.apply(this, arguments);
-            };
-          } else if (Array.isArray(preHandler)) {
-            opts[preHandlerKey].unshift(function prependedBeforeHandler(request, reply, done) {
-              annotateHttpEntrySpanWithPathTemplate(app, opts);
-              done();
-            });
-          }
-        }
-
-        return originalRoute.apply(this, arguments);
-      };
+    // NOTE: all major versions support `addHook` - this is just a safe protection
+    if (!app.addHook) {
+      logger.warn('Instana was not able to instrument Fastify. The instrumention of http requests is still working.');
+      return app;
     }
 
+    app.addHook('onRequest', function onRequest(request, reply, done) {
+      try {
+        // NOTE: v1 uses _context https://github.com/fastify/fastify/blob/1.x/fastify.js#L276
+        //       v2/v3 uses context https://github.com/fastify/fastify/blob/2.x/test/handler-context.test.js#L41
+        const url = reply._context ? reply._context.config.url : reply.context.config.url;
+
+        annotateHttpEntrySpanWithPathTemplate(app, { url });
```

Comment:
I usually prefer the option pattern even with single params to avoid that later on if there is another option I don't have to refactor the target function. But in this case, there might not be another cases for a second option. 

I'll change this 👍